include/deal.II/non_matching/fe_values.h

```diff
@@ -0,0 +1,419 @@
+             const MeshClassifier<dim> &  mesh_classifier,
+             const DoFHandler<dim> &      dof_handler,
+             const VectorType &           level_set,
+             const AdditionalData &       additional_data = AdditionalData());
+
+    /**
+     * Constructor.
+     *
+     * @param mapping_collection Collection of Mappings to be used.
+     * @param fe_collection Collection of FiniteElements to be used.
+     * @param q_collection Collection of Quadrature rules over $[0, 1]^{dim}$
+     * that should be used when a cell isn't intersected and we don't need to
+     * generate immersed quadrature rules.
+     * @param q_collection_1D Collection of 1-dimensional quadrature rules used
+     * to generate the immersed quadrature rules. See the QuadratureGenerator
+     * class.
+     * @param mesh_classifier Object used to determine when the immersed
+     * quadrature rules need to be generated.
+     * @param region_update_flags Struct storing UpdateFlags for the
+     * inside/outside/surface region of the cell.
+     * @param dof_handler The DoFHandler associated with the discrete level set
+     * function.
+     * @param level_set The degrees of freedom of the discrete level set function.
+     * @param additional_data Additional data passed to the QuadratureGenerator
+     * class.
+     *
+     * @note Pointers to @p mapping_collection, @p fe_collection,
+     * @p mesh_classifier, @p dof_handler, and @p level_set are stored
+     * internally, so these need to have a longer life span than the instance of
+     * this class.
+     */
+    template <class VectorType>
+    FEValues(const hp::MappingCollection<dim> &mapping_collection,
+             const hp::FECollection<dim> &     fe_collection,
+             const hp::QCollection<dim> &      q_collection,
+             const hp::QCollection<1> &        q_collection_1D,
+             const RegionUpdateFlags           region_update_flags,
+             const MeshClassifier<dim> &       mesh_classifier,
+             const DoFHandler<dim> &           dof_handler,
+             const VectorType &                level_set,
+             const AdditionalData &additional_data = AdditionalData());
+
+    /**
+     * Reinitialize the various FEValues-like objects for the 3 different
+     * regions of the cell. After calling this function an FEValues-like object
+     * can be retrieved for each region using the functions
+     * get_inside_fe_values(),
+     * get_outside_fe_values(), or
+     * get_surface_fe_values().
+     */
+    template <bool level_dof_access>
+    void
+    reinit(
+      const TriaIterator<DoFCellAccessor<dim, dim, level_dof_access>> &cell);
+
+    /**
+     * Return an dealii::FEValues object reinitialized with a quadrature for the
+     * inside region of the cell: $\{x \in K : \psi(x) < 0 \}$.
+     *
+     * @note If the quadrature rule over the region is empty the returned
+     * optional will not contain a value.
```

Comment:
```suggestion
     * @note If the quadrature rule over the region is empty, e.g. because the
     *  cell is completely located in the outside domain, the returned
     * optional will not contain a value.
```